src/output-json-dns.c

```diff
@@ -445,10 +479,241 @@ static void LogQuery(LogDnsLogThread *aft, json_t *js, DNSTransaction *tx,
+                if (hjs != NULL) {
+                   json_object_set_new(jdata, "sshfp", hjs);
+                }
+            }
+        }
+        json_array_append_new(js, jdata);
+    } while ((entry = TAILQ_NEXT(entry, next)));
+}
+
+static void OutputAnswerGrouped(DNSAnswerEntry *entry, json_t *js)
+{
+    struct {
+        #define ENTRY_TYPE_A        0
+        #define ENTRY_TYPE_AAAA     1
+        #define ENTRY_TYPE_TXT      2
+        #define ENTRY_TYPE_CNAME    3
+        #define ENTRY_TYPE_MX       4
+        #define ENTRY_TYPE_PTR      5
+        #define ENTRY_TYPE_NS       6
+        #define ENTRY_TYPE_SSHFP    7
+        #define ENTRY_TYPE_MAX      8
+        const char *name;
+        json_t *value;
+    } dns_rtypes[] = {
+        { "A",      NULL },
+        { "AAAA",   NULL },
+        { "TXT",    NULL },
+        { "CNAME",  NULL },
+        { "MX",     NULL },
+        { "PTR",    NULL },
+        { "NS",     NULL },
+        { "SSHFP",  NULL }
+    };
+
+    int i;
+    json_t *jrdata = json_object();
+    if (jrdata == NULL) {
+        return;
+    }
+
+    do {
+        uint8_t *ptr = (uint8_t *)((uint8_t *)entry + sizeof(DNSAnswerEntry)+ entry->fqdn_len);
+        if (entry->type == DNS_RECORD_TYPE_A && entry->data_len == 4) {
+            char a[16] = "";
+            if (dns_rtypes[ENTRY_TYPE_A].value == NULL) {
+                dns_rtypes[ENTRY_TYPE_A].value = json_array();
+                if (dns_rtypes[ENTRY_TYPE_A].value == NULL) {
+                    return;
+                }
+            }
+            PrintInet(AF_INET, (const void *)ptr, a, sizeof(a));
+            json_array_append_new(dns_rtypes[ENTRY_TYPE_A].value, json_string(a));
+        } else if (entry->type == DNS_RECORD_TYPE_AAAA && entry->data_len == 16) {
+            char a[46] = "";
+            if (dns_rtypes[ENTRY_TYPE_AAAA].value == NULL) {
+                dns_rtypes[ENTRY_TYPE_AAAA].value = json_array();
+                if (dns_rtypes[ENTRY_TYPE_AAAA].value == NULL) {
+                    return;
+                }
+            }
+            PrintInet(AF_INET6, (const void *)ptr, a, sizeof(a));
+            json_array_append_new(dns_rtypes[ENTRY_TYPE_AAAA].value, json_string(a));
+        } else if (entry->data_len == 0) {
+            json_object_set_new(js, "rdata", json_string(""));
+        } else if (entry->type == DNS_RECORD_TYPE_TXT || entry->type == DNS_RECORD_TYPE_CNAME ||
+                entry->type == DNS_RECORD_TYPE_MX || entry->type == DNS_RECORD_TYPE_PTR ||
+                entry->type == DNS_RECORD_TYPE_NS) {
+            if (entry->data_len != 0) {
+                char buffer[256] = "";
+                uint16_t copy_len = entry->data_len < (sizeof(buffer) - 1) ?
+                    entry->data_len : sizeof(buffer) - 1;
+                memcpy(buffer, ptr, copy_len);
+                buffer[copy_len] = '\0';
+
+                if (entry->type == DNS_RECORD_TYPE_TXT) {
```

Comment:
maybe this can be done with small util funcs or macros, to avoid the long list of very similar code